okio-zipfilesystem/src/jvmMain/kotlin/okio/zipfilesystem/ZipFileSystem.kt

```diff
@@ -0,0 +1,115 @@
+ * limitations under the License.
+ */
+package okio.zipfilesystem
+
+import okio.ExperimentalFileSystem
+import okio.FileMetadata
+import okio.FileSystem
+import okio.InflaterSource
+import okio.Path
+import okio.Path.Companion.toPath
+import okio.Sink
+import okio.Source
+import okio.buffer
+import java.io.FileNotFoundException
+import java.io.IOException
+import java.util.zip.Inflater
+
+/**
+ * Read only access to a zip file.
+ */
+@ExperimentalFileSystem
+class ZipFileSystem internal constructor(
+  private val zipPath: Path,
+  private val fileSystem: FileSystem,
+  private val entries: Map<Path, ZipEntry>,
+  private val comment: String?
+) : FileSystem() {
+  override fun canonicalize(path: Path): Path {
+    return "/".toPath() / path
+  }
+
+  override fun metadataOrNull(path: Path): FileMetadata? {
+    TODO()
+  }
+
+  override fun list(dir: Path): List<Path> {
+    // TODO: directories might be absent in the ZIP file.
+    val canonicalDir = canonicalize(dir)
+    return entries.keys.mapNotNull { if (it.parent == canonicalDir) it else null }
+  }
+
+  @Throws(IOException::class)
+  override fun source(path: Path): Source {
+    // Make sure this ZipEntry is in this Zip file.  We run it through the name lookup.
+    val canonicalPath = canonicalize(path)
+    val entry = entries[canonicalPath] ?: throw FileNotFoundException("no such file $path")
+    val source = fileSystem.source(zipPath).buffer()
+    val cursor = source.cursor()!!
+
+    // We don't know the entry data's start position. All we have is the
+    // position of the entry's local header.
+    // http://www.pkware.com/documents/casestudies/APPNOTE.TXT
+    cursor.seek(entry.localHeaderRelOffset)
+
+    val localMagic = source.readIntLe()
+    if (localMagic.toLong() != LOCSIG) {
+      throwZipException("Local File Header", localMagic)
+    }
+    source.skip(2)
+
+    // At position 6 we find the General Purpose Bit Flag.
+    val gpbf = source.readShortLe().toInt() and 0xffff
+    if (gpbf and GPBF_UNSUPPORTED_MASK != 0) {
+      throw IOException("Invalid General Purpose Bit Flag: $gpbf")
+    }
+
+    // Offset 26 has the file name length, and offset 28 has the extra field length.
+    // These lengths can differ from the ones in the central header.
+    source.skip(18)
+    val fileNameLength = source.readShortLe().toInt() and 0xffff
+    val extraFieldLength = source.readShortLe().toInt() and 0xffff
+    // Skip the variable-size file name and extra field data.
+    source.skip((fileNameLength + extraFieldLength).toLong())
+
+    return when (entry.compressionMethod) {
+      ZipEntry.STORED -> {
+        FixedLengthSource(source, entry.size, truncate = true)
+      }
+      else -> {
+        val inflaterSource = InflaterSource(
+          FixedLengthSource(source, entry.compressedSize, truncate = true),
+          Inflater(true)
+        )
+        FixedLengthSource(inflaterSource, entry.size, truncate = false)
```

Comment:
I understand about truncated the compressed size so you don't overrun, but why truncate = false here?